File: Problem_Sets/LeetCode/FirstBadVersion/FirstBadVersion.cpp
// The API isBadVersion is defined for you.
// bool isBadVersion(int version);

class Solution {
public:
    int firstBadVersion(int n) {
        int first = 1;
        int last = n;
        // Records 1st index of faulty product
        long pos = 0;
        while (first <= last) {
            int mid = first + (last - first) / 2;
            bool isBad = isBadVersion(mid);
            if (isBad) {
                pos = mid;
                last = mid - 1;
            } else {
                first = mid + 1;
            }
        }
        return pos;
    }
};

File: Problem_Sets/LeetCode/FirstBadVersion/FirstBadVersion.py
# The isBadVersion API is already defined for you.
# def isBadVersion(version: int) -> bool:

class Solution:
    def firstBadVersion(self, n: int) -> int:
        first, last, min = 1, n, 0
        while first <= last:
            mid = first + (last - first) // 2
            isBad = isBadVersion(mid)
            if isBad:
                min = mid
                last = mid - 1
            else:
                first = mid + 1
        return min
        